src/Model/CodecCursor.php

```diff
@@ -0,0 +1,126 @@
+<?php
+/*
+ * Copyright 2023-present MongoDB, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+namespace MongoDB\Model;
+
+use Iterator;
+use MongoDB\BSON\Document;
+use MongoDB\Codec\DocumentCodec;
+use MongoDB\Driver\Cursor;
+use MongoDB\Driver\CursorId;
+use MongoDB\Driver\CursorInterface;
+use MongoDB\Driver\Server;
+
+use function assert;
+use function iterator_to_array;
+
+/**
+ * @template TValue of object
+ * @template-implements CursorInterface<int, TValue>
+ * @template-implements Iterator<int, TValue>
+ */
+class CodecCursor implements CursorInterface, Iterator
+{
+    private const TYPEMAP = ['root' => 'bson'];
+
+    private Cursor $cursor;
+
+    /** @var DocumentCodec<TValue> */
+    private DocumentCodec $codec;
+
+    /** @var TValue|null */
+    private ?object $current = null;
+
+    /** @return TValue */
+    public function current(): ?object
+    {
+        if (! $this->current && $this->valid()) {
+            $value = $this->cursor->current();
+            assert($value instanceof Document);
+            $this->current = $this->codec->decode($value);
+        }
+
+        return $this->current;
+    }
+
+    /**
+     * @template NativeClass of Object
+     * @param DocumentCodec<NativeClass> $codec
+     * @return self<NativeClass>
+     */
+    public static function fromCursor(Cursor $cursor, DocumentCodec $codec): self
+    {
+        $cursor->setTypeMap(self::TYPEMAP);
+
+        return new self($cursor, $codec);
+    }
+
+    public function getId(): CursorId
+    {
+        return $this->cursor->getId();
+    }
+
+    public function getServer(): Server
+    {
+        return $this->cursor->getServer();
+    }
+
+    public function isDead(): bool
+    {
+        return $this->cursor->isDead();
+    }
+
+    public function key(): int
+    {
+        return $this->cursor->key();
+    }
+
+    public function next(): void
+    {
+        $this->current = null;
+        $this->cursor->next();
+    }
+
+    public function rewind(): void
+    {
+        $this->current = null;
+        $this->cursor->rewind();
+    }
+
+    public function setTypeMap(array $typemap): void
+    {
+        // Not supported
```

Comment:
I decided to make this a no-op, as throwing an exception would most violate the Liskov substitution principle.

Reply:
Would `E_USER_WARNING` be appropriate? If this is a new class that only becomes relevant when using codecs, I'm curious how users would hit this unless they had already modified their application to use codecs. In that sense, this doesn't seem much different than changing the return type elsewhere from `Cursor` to `CursorInterface&Iterator` and not worrying about the theoretical (but impossible in practice?) BC break.